compiler/resolution/preFold.cpp

```diff
@@ -390,59 +395,55 @@ static void setRecordDefaultValueFlags(AggregateType* at) {
       ts->addFlag(FLAG_TYPE_NO_DEFAULT_VALUE);
     } else if (isNilableClassType(at)) {
       ts->addFlag(FLAG_TYPE_DEFAULT_VALUE);
-
     } else if (Type* eltType = arrayElementType(at)) {
-      if (AggregateType* eltTypeAt = toAggregateType(eltType)) {
-        setRecordAssignableFlags(eltTypeAt);
-
-        if (eltType->symbol->hasFlag(FLAG_TYPE_DEFAULT_VALUE))
-          at->symbol->addFlag(FLAG_TYPE_DEFAULT_VALUE);
-        else if (eltType->symbol->hasFlag(FLAG_TYPE_NO_DEFAULT_VALUE))
-          at->symbol->addFlag(FLAG_TYPE_NO_DEFAULT_VALUE);
+      if (isDefaultInitializable(eltType)) {
+        ts->addFlag(FLAG_TYPE_DEFAULT_VALUE);
       } else {
-         at->symbol->addFlag(FLAG_TYPE_DEFAULT_VALUE);
+        ts->addFlag(FLAG_TYPE_NO_DEFAULT_VALUE);
       }
 
-
     } else if (at->symbol->hasFlag(FLAG_EXTERN)) {
       // Currently extern records aren't initialized at all by default.
       // But it's not necessarily reasonable to expect them to have
       // initializers. See issue #7992.
       ts->addFlag(FLAG_TYPE_DEFAULT_VALUE);
     } else {
+
+      if (at->hasUserDefinedInit == false ||
+          at->symbol->hasFlag(FLAG_TUPLE)) {
+        // For records with compiler-generated default init,
+        // or for tuples, first consider the fields.
+        // If any of the fields prohibit
+        // default initialization, return before trying
+        // to resolve the init() function.
+        bool failsDefaultInit = false;
+        for_fields(field, at) {
+          Type* fieldType = field->getValType(); // val type for tuples
+          if (isDefaultInitializable(fieldType) == false) {
+            // check for default value
+            if (field->defPoint->init == NULL) {
```

Comment:
Can you explain why you have to check for the init function even `isDefaultInitializable == false`?

Reply:
This is not checking for an init function. It is checking for a initialization expression on the field.

Reply:
Thanks!